[1.19] Fix several bugs in the original grindstone PR related to canceling the event

KnightMiner opened the pull request:
Fixes issues related to canceling the grindstone event. Includes test mod from #8578

Fix canceling the event not clearing the result item, causing it to dupe whatever the result with the previous item was
Fix canceling the event running vanilla behavior if vanilla behavior ran at all since the menu was opened and no other listener changed the XP. This also fixes the grindstone event behavior disagreeing with the javadoc (javadoc said XP will be -1 for first consumer, but instead XP was whatever the last XP was)

KnightMiner committed:
Fix several bugs in the original PR related to canceling the event

Fix canceling the event not clearing the result item, causing it to dupe whatever the result with the previous item was
Fix canceling the event running vanilla behavior if vanilla behavior ran at all since the menu was opened and no other listener changed the XP

TheCurle commented:
Please reword the title of the PR: what's the "original"? #1?

ferriarnus commented:
I'll apologize about the mistakes I've made in the previous pr. I can't get to my pc so I'm not to sure what I did with the javadoc, but from what I can recall, it should take the xp and not -1, so when the event is called multiple times, you have the latest xp that would be present. But I might have messed that up.

KnightMiner commented:
The problem with taking the last XP is it's never cleared, causing the behavior to be inconsistent between if you place items in the container right after it opens vs place items when it's been open for awhile. It's not like it's the real XP value, it's from a previous instance of the event, so whatever code you have setting XP the first time should set it just fine the second time

KnightMiner committed:
Add comments to the test mod explaining the expected behavior

KnightMiner committed:
Clarify docs on behavior of the XP

Makes it clear XP is ignored in on place if the output is empty or the event is canceled

KnightMiner committed:
Fix broadcast changes not being called if the event overrides grindstone behavior

Not sure its strictly needed, but every vanilla branch in the function calls it and its a trivial change.

Matyrobbrt commented on src/test/java/net/minecraftforge/debug/misc/GrindstoneEventTest.java:
```diff
@@ -0,0 +1,79 @@
+
+@Mod("grindstone_event_test")
+@Mod.EventBusSubscriber
+public class GrindstoneEventTest {
```
Please add an `ENABLED` field, defaulted to `false`, which disables the event listeners (manually add the listeners using `MinecraftForge.EVENT_BUS#addListener`/`MinecraftForge.EVENT_BUS#register`)

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -67,8 +67,8 @@ public void setXp(int xp)
      * It is called from {@link GrindstoneMenu#createResult()}. <br>
      * If the event is canceled, vanilla behavior will not run, and the output will be set to {@link ItemStack#EMPTY}. <br>
```
I would recommend the following Javadoc:
```java
/**
 * This event is {@link Cancelable} <br>
 * {@link OnplaceItem} is fired from {@link GrindstoneMenu#createResult()} when the inputs of a grindstone are changed. <br>
 *
 * The following rules apply:
 * <ul>
 *     <li>If the event is canceled, vanilla behavior will not run, and the output will be {@linkplain ItemStack#EMPTY empty}.</li>
 *     <br>
 *     <li>If the event is not canceled</li>
 *     <ul>
 *          <li>and the output is empty, the output will be determined by vanilla.</li>
 *          <li>and the output is not empty, the output will be set, without running vanilla behavior.</li>
 *     </ul>
 *     <br>
 *     <li>Vanilla XP calculation logic will be used unless all of the following criterias are met:</li>
 *     <ul>
 *         <li>the amount of experience is greater than or equal to {@code 0};</li>
 *         <li>the event is not {@linkplain #isCanceled() canceled};</li>
 *         <li>the {@linkplain #getOutput() output} is not empty.</li>
 *     </ul>
 * </ul>
 */
```

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -41,10 +41,10 @@ public ItemStack getBottomItem()
-     * If you are the first receiver of this event, it is guaranteed to be -1. <br>
-     * The value must be set to a positive value to override vanilla behavior. <br>
-     * if the value is equal to -1, the vanilla behavior for calculating experience will run. <br>
+     * This is the experience amount determined by the event, it will be -1 unless a mod calls {@link #setXp(int)}. <br>
```
```suggestion
     * This is the experience amount determined by the event. It will be {@code -1} unless {@link #setXp(int)} is called. <br>
```

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -41,10 +41,10 @@ public ItemStack getBottomItem()
-     * The value must be set to a positive value to override vanilla behavior. <br>
-     * if the value is equal to -1, the vanilla behavior for calculating experience will run. <br>
+     * This is the experience amount determined by the event, it will be -1 unless a mod calls {@link #setXp(int)}. <br>
+     * If set to a non-negative value, that value will be used instead of running vanilla behavior. <br>
+     * If the value is negative, the vanilla behavior for calculating experience will run. <br>
+     * Ignored if the output is empty or the event is canceled.
```
Can be removed, as that is already documented in the event Javadoc.

Matyrobbrt commented on src/main/java/net/minecraftforge/common/ForgeHooks.java:
```diff
@@ -759,7 +759,11 @@ public static float onAnvilRepair(Player player, @NotNull ItemStack output, @Not
+        if (MinecraftForge.EVENT_BUS.post(e))
+        {
+            outputSlot.setItem(0, ItemStack.EMPTY);
+            return -1;
```
Return the `xp` parameter instead (which will be `-1` if fired by vanilla).

KnightMiner replied:
That would be wrong. XP might have been set before it was cancelled, we want to be certain the result is a known value and not one of our technical values 

The value is entirely ignored if the event is cancelled, might as well make it predictable

Matyrobbrt reviewed (changes requested):
A few Javadoc changes

KnightMiner committed:
Add enabled flag to the test mod

KnightMiner committed:
Improve javadocs

Matyrobbrt commented on src/main/java/net/minecraftforge/event/GrindstoneEvent.java:
```diff
@@ -64,11 +61,22 @@ public void setXp(int xp)
     /**
      * This event is {@link Cancelable} <br>
      * {@link OnplaceItem} is fired when the inputs to a grindstone are changed. <br>
```
While you're changing the javadoc, might as well replace this `to` with a `of`.
